lib/src/configuration.dart

```diff
@@ -395,21 +418,315 @@ class RealmSchema extends Iterable<SchemaObject> {
+  /// are properly propagated through the standard Realm's change notifications.
+  /// This strategy supplies four callbacks: [onBeforeReset], [onAfterRecovery], [onAfterDiscard] and [onManualResetFallback].
+  /// [onBeforeReset] is invoked just before the client reset happens.
+  /// [onAfterRecovery] is invoke if and only if an automatic client reset succeeded. The callback is never called
+  /// if the automatic client reset fails.
+  /// [onAfterDiscard] is invoked if and only if an automatic client reset failed and instead the discard unsynced one succeded.
+  /// The callback is never called if the discard unsynced client reset fails.
+  /// [onManualResetFallback] is invoked whenever an error occurs in either of the recovery stragegies and the system needs to fallback to a manual mode.
+  const RecoverOrDiscardUnsyncedChangesHandler(
+      {BeforeResetCallback? onBeforeReset, AfterResetCallback? onAfterRecovery, AfterResetCallback? onAfterDiscard, ClientResetCallback? onManualResetFallback})
+      : super._(ClientResyncModeInternal.recoverOrDiscard, onManualResetFallback,
+            onBeforeReset: onBeforeReset, onAfterDiscard: onAfterDiscard, onAfterRecovery: onAfterRecovery);
+}
 
 /// @nodoc
-extension SchemaObjectInternal on SchemaObject {
-  bool get isGenericRealmObject => type == RealmObject || type == EmbeddedObject || type == RealmObjectBase;
+extension ClientResetHandlerInternal on ClientResetHandler {
+  ClientResyncModeInternal get clientResyncMode => _mode;
+
+  BeforeResetCallback? get onBeforeReset => _onBeforeReset;
+  AfterResetCallback? get onAfterDiscard => _onAfterDiscard;
+  AfterResetCallback? get onAfterRecovery => _onAfterRecovery;
+}
+
+/// Enum describing what should happen in case of a Client Resync.
+/// @nodoc
+enum ClientResyncModeInternal {
+  manual,
+  discardLocal,
+  recover,
+  recoverOrDiscard,
+}
+
+/// An error type that describes a client reset error condition.
+/// {@category Sync}
+class ClientResetError extends SyncError {
+  /// If true the received error is fatal.
+  final bool isFatal = true;
+  final Configuration? config;
+
+  /// The [ClientResetError] has error code of [SyncClientErrorCode.autoClientResetFailure]
+  SyncClientErrorCode get code => SyncClientErrorCode.autoClientResetFailure;
+
+  ClientResetError(String message, [this.config]) : super(message, SyncErrorCategory.client, SyncClientErrorCode.autoClientResetFailure.code);
+
+  @override
+  String toString() {
+    return "SyncError message: $message category: $category code: $code isFatal: $isFatal";
+  }
+
+  /// Initiates the client reset process.
+  ///
+  /// All Realm instances for that path must be closed before this method is called or an
+  /// Exception will be thrown.
+  void resetRealm() {
+    if (config == null || config is! FlexibleSyncConfiguration) {
+      throw Exception("FlexibleSyncConfiguration is not set into ClientResetError");
```

Comment:
All exception that the SDK throws should be RealmExceptions or inheritors. This is also true for errors where it should be RealmError

Reply:
also the exception text is not clear. What are we trying to say here?

Probably `The current configuration is not FlexibleSyncConfiguration`